File: trackers/tcs.py
from datetime import datetime
from .base import BaseTracker


class TCSTracker(BaseTracker):
    """Tracker for TCS Express parcels."""

    API_URL = "https://www.tcsexpress.com/apibridge"

    def track(self, tracking_number):
        """Track a TCS parcel using their API."""
        try:
            headers = {
                'Content-Type': 'application/json; charset=UTF-8',
                'Origin': 'https://www.tcsexpress.com',
                'Referer': f'https://www.tcsexpress.com/track/{tracking_number}',
            }

            # TCS uses apibridge with specific format
            # Headers field encodes the tracking number as characters
            headers_dict = {}
            for i, char in enumerate(tracking_number):
                headers_dict[str(i)] = char

            payload = {
                "body": {
                    "url": "trackapinew",
                    "type": "GET",
                    "headers": headers_dict,
                    "payload": {},
                    "param": f"consignee={tracking_number}"
                }
            }

            resp = self.session.post(
                self.API_URL,
                json=payload,
                headers=headers,
                timeout=30
            )
            resp.raise_for_status()
            data = resp.json()

            if not data.get('isSuccess'):
                msg = data.get('responseData', 'Unknown error')
                return {"error": f"API error: {msg}"}

            response_data = data.get('responseData', {})

            # Check for no data
            if isinstance(response_data, str):
                if 'No Data Found' in response_data or 'Invalid' in response_data:
                    return {"error": f"No data or invalid tracking: {tracking_number}"}
                return {"error": f"API error: {response_data}"}

            shipment_info_raw = response_data.get('shipmentinfo')
            shipment_info = shipment_info_raw[0] if isinstance(shipment_info_raw, list) and shipment_info_raw else shipment_info_raw or {}
            delivery_info_raw = response_data.get('deliveryinfo')
            delivery_info = delivery_info_raw[0] if isinstance(delivery_info_raw, list) and delivery_info_raw else delivery_info_raw or {}
            checkpoints = response_data.get('checkpoints') or []

            # Extract status
            status = delivery_info.get('currentStatus', '') or \
                     shipment_info.get('status', '') or \
                     response_data.get('shipmentsummary', 'Unknown')

            if not status or status == 'Unknown':
                # Try to get status from checkpoints
                if checkpoints:
                    status = checkpoints[0].get('status', 'Unknown')

            # Extract location
            location = delivery_info.get('currentLocation', '') or \
                       shipment_info.get('destination', 'N/A')

            # Parse history from checkpoints
            history = []
            for cp in checkpoints:
                history.append({
                    'timestamp': cp.get('datetime', cp.get('statusDate', '')),
                    'status': cp.get('status', ''),
                    'location': cp.get('location', '')
                })

            delivered = 'delivered' in str(status).lower() or 'received' in str(status).lower()

            return {
                'status': status,
                'location': location,
                'timestamp': datetime.now().isoformat(),
                'history': history,
                'delivered': delivered
            }

        except Exception as e:
            return {"error": f"Tracking failed: {str(e)}"}

File: src/server/trackers/tcs.ts
import type { TrackingResult, HistoryItem } from './types.ts';

export async function trackTCS(trackingNumber: string): Promise<TrackingResult> {
  const url = "https://www.tcsexpress.com/apibridge";
  try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 20000);

    const headersDict: Record<string, string> = {};
    for (let i = 0; i < trackingNumber.length; i++) {
      headersDict[String(i)] = trackingNumber[i];
    }

    const payload = {
      body: {
        url: "trackapinew",
        type: "GET",
        headers: headersDict,
        payload: {},
        param: `consignee=${trackingNumber}`
      }
    };

    const resp = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json; charset=UTF-8",
        "Origin": "https://www.tcsexpress.com",
        "Referer": `https://www.tcsexpress.com/track/${encodeURIComponent(trackingNumber)}`,
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
      },
      body: JSON.stringify(payload),
      signal: controller.signal
    });
    clearTimeout(timeout);

    if (!resp.ok) {
      return {
        status: "Error",
        location: "N/A",
        timestamp: new Date().toISOString(),
        history: [],
        delivered: false,
        error: `HTTP ${resp.status} ${resp.statusText}`
      };
    }

    const data = await resp.json() as any;

    if (!data?.isSuccess) {
      const msg = data?.responseData || "Unknown error";
      return {
        status: "Error",
        location: "N/A",
        timestamp: new Date().toISOString(),
        history: [],
        delivered: false,
        error: `API error: ${typeof msg === 'string' ? msg : JSON.stringify(msg)}`
      };
    }

    const responseData = data.responseData || {};
    if (typeof responseData === "string") {
      if (responseData.includes("No Data Found") || responseData.includes("Invalid")) {
        return {
          status: "Error",
          location: "N/A",
          timestamp: new Date().toISOString(),
          history: [],
          delivered: false,
          error: `No data or invalid tracking: ${trackingNumber}`
        };
      }
      return {
        status: "Error",
        location: "N/A",
        timestamp: new Date().toISOString(),
        history: [],
        delivered: false,
        error: `API error: ${responseData}`
      };
    }

    const shipmentInfoRaw = responseData.shipmentinfo;
    const shipmentInfo = Array.isArray(shipmentInfoRaw) ? shipmentInfoRaw[0] : (shipmentInfoRaw || {});
    const deliveryInfoRaw = responseData.deliveryinfo;
    const deliveryInfo = Array.isArray(deliveryInfoRaw) ? deliveryInfoRaw[0] : (deliveryInfoRaw || {});
    const checkpoints = Array.isArray(responseData.checkpoints) ? responseData.checkpoints : [];

    let status = deliveryInfo.currentStatus || shipmentInfo.status || "";
    if (!status || status === "Unknown") {
      if (checkpoints.length > 0 && checkpoints[0].status) {
        status = checkpoints[0].status;
      } else {
        const rawSummary = responseData.shipmentsummary || "Unknown";
        if (rawSummary.includes("No Data Found") || rawSummary.includes("Invalid")) {
          status = "Awaiting First Scan (No Data Found yet)";
        } else {
          status = rawSummary;
        }
      }
    } else if (status.includes("No Data Found") || status.includes("Invalid")) {
      status = "Awaiting First Scan (No Data Found yet)";
    }

    let location = deliveryInfo.currentLocation || shipmentInfo.destination || "";
    const history: HistoryItem[] = [];
    for (const cp of checkpoints) {
      history.push({
        timestamp: cp.datetime || cp.statusDate || "",
        status: cp.status || "",
        location: cp.recievedby || ""
      });
    }

    if (!location || location === "N/A") {
      for (const cp of checkpoints) {
        if (cp.recievedby) {
          location = cp.recievedby;
          break;
        }
      }
    }
    if (!location) location = "N/A";

    const statusLower = String(status).toLowerCase();
    const delivered = statusLower.includes("delivered") || statusLower.includes("received");

    return {
      status,
      location,
      timestamp: new Date().toISOString(),
      history,
      delivered
    };
  } catch (err: any) {
    return {
      status: "Error",
      location: "N/A",
      timestamp: new Date().toISOString(),
      history: [],
      delivered: false,
      error: `Tracking failed: ${err.message || String(err)}`
    };
  }
}
